### ISSUE TITLE
Strange log information during training

### ISSUE THREAD

**Issue by qsisi** [ShengyuH/OverlapPredator.Mink#1]
Hello,

Thank you for posting the code based on MinkowskiEngine,

I ran the training script:
`python main.py configs/train/indoor.yaml`
after several epoches, It comes out with some warnings after every training and validation process:
**UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.**
**UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 due to no true samples. Use `zero_division` parameter to control this behavior.**

and I check the log file in `./snapshot/indoor/log`, which looks like the following:

**#parameters 10.860164 M
#parameters 10.860164 M
#parameters 10.860164 M
train Epoch: 1 [ 500/2580]circle_loss: 0.00	recall: 0.00	saliency_loss: 0.00	saliency_recall: 0.00	saliency_precision: 0.00	overlap_loss: 0.00	overlap_recall: 0.00	overlap_precision: 0.00	
train Epoch: 1 [1000/2580]circle_loss: 0.00	recall: 0.00	saliency_loss: 0.00	saliency_recall: 0.00	saliency_precision: 0.00	overlap_loss: 0.00	overlap_recall: 0.00	overlap_precision: 0.00	
train Epoch: 1 [1500/2580]circle_loss: 0.00	recall: 0.00	saliency_loss: 0.00	saliency_recall: 0.00	saliency_precision: 0.00	overlap_loss: 0.00	overlap_recall: 0.00	overlap_precision: 0.00	
train Epoch: 1 [2000/2580]circle_loss: 0.00	recall: 0.00	saliency_loss: 0.00	saliency_recall: 0.00	saliency_precision: 0.00	overlap_loss: 0.00	overlap_recall: 0.00	overlap_precision: 0.00	
train Epoch: 1 [2500/2580]circle_loss: 0.00	recall: 0.00	saliency_loss: 0.00	saliency_recall: 0.00	saliency_precision: 0.00	overlap_loss: 0.00	overlap_recall: 0.00	overlap_precision: 0.00	
train Epoch: 1circle_loss: 0.00	recall: 0.00	saliency_loss: 0.00	saliency_recall: 0.00	saliency_precision: 0.00	overlap_loss: 0.00	overlap_recall: 0.00	overlap_precision: 0.00	
val Epoch: 1circle_loss: 1.50	recall: 0.01	saliency_loss: 0.00	saliency_recall: 0.34	saliency_precision: 0.00	overlap_loss: 0.33	overlap_recall: 0.28	overlap_precision: 0.41	
Save model to snapshot/indoor/checkpoints/model_best_loss.pth
Save model to snapshot/indoor/checkpoints/model_best_recall.pth
train Epoch: 2 [ 500/2580]circle_loss: 0.00	recall: 0.00	saliency_loss: 0.00	saliency_recall: 0.00	saliency_precision: 0.00	overlap_loss: 0.00	overlap_recall: 0.00	overlap_precision: 0.00	
train Epoch: 2 [1000/2580]circle_loss: 0.00	recall: 0.00	saliency_loss: 0.00	saliency_recall: 0.00	saliency_precision: 0.00	overlap_loss: 0.00	overlap_recall: 0.00	overlap_precision: 0.00	
train Epoch: 2 [1500/2580]circle_loss: 0.00	recall: 0.00	saliency_loss: 0.00	saliency_recall: 0.00	saliency_precision: 0.00	overlap_loss: 0.00	overlap_recall: 0.00	overlap_precision: 0.00	
train Epoch: 2 [2000/2580]circle_loss: 0.00	recall: 0.00	saliency_loss: 0.00	saliency_recall: 0.00	saliency_precision: 0.00	overlap_loss: 0.00	overlap_recall: 0.00	overlap_precision: 0.00	
train Epoch: 2 [2500/2580]circle_loss: 0.00	recall: 0.00	saliency_loss: 0.00	saliency_recall: 0.00	saliency_precision: 0.00	overlap_loss: 0.00	overlap_recall: 0.00	overlap_precision: 0.00	
train Epoch: 2circle_loss: 0.00	recall: 0.00	saliency_loss: 0.00	saliency_recall: 0.00	saliency_precision: 0.00	overlap_loss: 0.00	overlap_recall: 0.00	overlap_precision: 0.00	
val Epoch: 2circle_loss: 1.49	recall: 0.01	saliency_loss: 0.00	saliency_recall: 0.33	saliency_precision: 0.00	overlap_loss: 0.33	overlap_recall: 0.28	overlap_precision: 0.40	
Save model to snapshot/indoor/checkpoints/model_best_loss.pth**

does the above warnings account for all the zeros in the printout information? 
Also, after 20 epoches, the overall loss doesn't look like it's going down, which stablize at around 1.48, the best_recall is 0.0149.

Thank you very much for your help.

**Comment by ShengyuH** [comment]
hi, 

I don't think the above warnings account for all-zeros in the log. I just quickly tested and this is the log: 

```
#parameters 10.860164 M
train Epoch: 1 [ 500/20642]circle_loss: 1.35	recall: 0.01	saliency_loss: 0.01	saliency_recall: 0.45	saliency_precision: 0.00	overlap_loss: 0.29	overlap_recall: 0.54	overlap_precision: 0.46	
```

Please see assets/log for full log information. 

Please make sure that you are
- working on the latest codebase
- using the training data provided by us
- having the same MinkowskiEngine and PyTorch version
- using batch_size = 1. Current implementation doesn't support batch processing. 
 
In addition, you can use 
```
import pdb
pdb.set_trace()
```
to debug the intermediate results and see where things go wrong.

Let me know if this helps. 

Best,
Shengyu

**Comment by qsisi** [comment]
Thanks for the timely reply,

after setting the batch_size = 1, it goes well. I guess setting the batch_size > 1 means using multi-pointclouds as a "whole" input cloud which will mislead the "CrossAttention" block ? 

Anyway, Thanks for your help.

**Comment by qsisi** [comment]
Hello, I have several questions about the Table 6 in your paper:

1. The sample number for each scene in 3DMatch test dataset is smaller than the original 3DMatch test dataset, is it because you remove the consecutive pairs in test set? 
2. What do the sample number '160' and '128' below Avg. and STD mean? 
3. Looks like in your code you compute the **mean-precision** and **weighted-preicision** but no **mean-recall**, does that mean the **Avg Registration Recall** in Table 6 is actually **mean-precision** but not recall?

Sorry to bother you again, and thank you very much for your help.

**Comment by ShengyuH** [comment]
hi,

> 1. The sample number for each scene in 3DMatch test dataset is smaller than the original 3DMatch test dataset, is it because you remove the consecutive pairs in test set?
---Yes

> 2. What do the sample number '160' and '128' below Avg. and STD mean?
--- I don't understand your question. This two numbers just tell us that the numbers of test pairs per scene vary a lot, the equally-weighted average mean might be contaminated by gross errors, though it turns out that this is not the case. 

> 3. Looks like in your code you compute the **mean-precision** and **weighted-preicision** but no **mean-recall**, does that mean the **Avg Registration Recall** in Table 6 is actually **mean-precision** but not recall?
--- YES, we follow other works(e.g., D3Feat) here. We estimate poses for all given pairs, and then directly check the success rate. Since we are not missing a single pair, this can also be interpreted as "recall". Please have a look at here: https://github.com/zgojcic/3D_multiview_reg/blob/22468adadfccaffe92c6d88a4cce42b5b7edb77e/lib/utils.py#L584

**Comment by qsisi** [comment]
Thanks for your timely reply! Now I totally catch your point.

**Comment by qsisi** [comment]
Hello! 

Sorry to bother you again, I want to make sure which metric did you report in **Table 2** in your paper? **Mean Precision** or **Weighted Precision**? As far as I know, two precisions above denotes **scene-level** and **pair-level** recall respectively, it's that right?

Thank you very much for your help.

**Comment by ShengyuH** [comment]
I report mean precision. Your understanding is correct.

**Comment by qsisi** [comment]
Thanks!
